File: Miet/divD/Training-5/E1.cpp
/**
 *    author:  kostya
 *    created: 2021-08-01 22:33:10
 *    modified 2021-08-02 00:10:36
 **/

#include <utility>
#include <vector>
#include <iostream>
#include <string>
#include <algorithm>

class Command {
public:
		std::string	name;
		int			cnt;
		int			score;

		Command (std::string  _name, int _cnt, int _score)
				: name(std::move(_name)), cnt(_cnt), score(_score) {}

		[[nodiscard]]bool operator ==(int other) const {
			return (other == score);
		}
};

template <typename _Bidirect_Iter_Tp, typename _Cmp_Tp>
int cmd_count(_Bidirect_Iter_Tp __first, const _Bidirect_Iter_Tp &__last, _Cmp_Tp __cmp) {
	_Cmp_Tp cnt = {};
	while (__first != __last) {
		if (*__first == __cmp)
			++cnt;
		++__first;
	}
	return cnt;
}

int main() {
	int n, k;
	std::cin >> n >> k;
	std::vector<std::string> cmd_names(n);
	std::vector<int> cmd_ans_cnt(n);
	std::vector<int> cmd_scores(n);
	std::vector<int> qu_scores(n, 1);
	std::vector<std::vector<bool>>ans_mat(n);
	char ans;

	for (int i=0; i < n; ++i) {
		std::cin >> cmd_names[i];
		ans_mat[i].resize(k);
		for (int j=0; j < k; ++j) {
			std::cin >> ans;
			cmd_ans_cnt[i] += ans == '+';
			qu_scores[j] += ans == '-';
			ans_mat[i][j] = ans == '+';
		}
	}
	for (int i=0; i < n; ++i) {
		for (int j=0; i < k; ++j) {
			cmd_scores[i] += qu_scores[j] * ans_mat[i][j];
		}
	}

	std::vector<Command> Commands;
	for (int i=0; i < n; ++i) {
		Commands.emplace_back(Command(cmd_names[i], cmd_ans_cnt[i], cmd_scores[i]));
	}

	std::vector<std::vector<Command>> commands_bins;
	for (int i=0; i < n; ++i) {
		commands_bins[cmd_ans_cnt[i]].emplace_back(Commands[i]);
	}

	int place = 1;
	for (const auto &cmd_bin_ : commands_bins) {
		std::sort(cmd_bin_.begin(), cmd_bin_.end(), [](const Command &_Cmd){return _Cmd.name;});
		std::sort(cmd_bin_.begin(), cmd_bin_.end(), [](const Command &_Cmd){return _Cmd.score;});
		int cmd_ = 0;
		while (cmd_ < cmd_bin_.size()) {
			int cnt = cmd_count(cmd_bin_.begin(), cmd_bin_.end(), cmd_bin_[cmd_].score);
			if (cnt == 1) {
				cout << place << ' ' << cmd_bin_[cmd_].name << endl;;
				++cmd_;
			} else {
				for (int place_tmp_ = 0; place_tmp_ < cnt; ++place_tmp_) {
					cout << place << '-' << place + cnt - 1 << ' ' << cmd_bin_[cmd_].name << endl;
				}
			}
			place += cnt;
		}
	}
}

File: Miet/divD/Training-5/E1.py
##
#    author:  kostya
#    created: 2021-08-01 16:59:38
#    modified 2021-08-01 23:19:12
##

class Command(object):
	"""
	"""
	def __init__(self, _name, _cnt, _sc):
		self.name = _name
		self.cnt = _cnt
		self.score = _sc

	def __eq__(self, other):
		return self.score == other

	def __repr__(self):
		return '{' + self.name + ', ' + str(self.cnt) + ', ' + str(self.score) + '}'


n, k = map(int, input().split())
cmd_names = []
cmd_ans_cnt = [0] * n
cmd_scores = [0] * n
qu_scores = [1] * k
ans_mat = [[0] * k for i in range(n)]

for i in range(n):
	cmd_name_, *ans_ar_ = input().split()
	cmd_names.append(cmd_name_)
	for j in range(k):
		cmd_ans_cnt[i] += ans_ar_[j] == '+'
		qu_scores[j] += ans_ar_[j] == '-'
		ans_mat[i][j] = ans_ar_[j] == '+'
for i in range(n):
	for j in range(k):
		cmd_scores[i] += qu_scores[j] * ans_mat[i][j]

Commands = []
for i in range(n):
	Commands.append(Command(cmd_names[i], cmd_ans_cnt[i], cmd_scores[i]))

commands_bins = [[] for _i in range(k + 1)]
for i in range(n):
	commands_bins[cmd_ans_cnt[i]].append(Commands[i])
place = 1
for cmd_bin_ in commands_bins[::-1]:
	cmd_bin_.sort(key=lambda _Cmd: _Cmd.name)
	cmd_bin_.sort(key=lambda _Cmd: _Cmd.score, reverse=True)
	cmd_ = 0
	while cmd_ < len(cmd_bin_):
		cnt = cmd_bin_.count(cmd_bin_[cmd_].score)
		if cnt == 1:
			place_str = str(place)
			print(place_str, cmd_bin_[cmd_].name)
			cmd_ += 1
		else:
			for place_tmp in range(cnt):
				place_str = str(place) + '-' + str(place + cnt - 1)
				print(place_str, cmd_bin_[cmd_].name)
				cmd_ += 1
		place += cnt

